native/common/include/jp_array.h

```diff
@@ -25,13 +25,34 @@ struct _object;
 typedef _object PyObject;
 #endif
 
+class JPArray;
+
+class JPArrayView
+{
+public:
+	JPArrayView(JPArray* array);
```

Comment:
Wouldn't the usage of a smart pointer protect us from releasing the JPArray too early? I guess you have dealt with this problem elsewhere.

Reply:
Arrayviews must be reference counted or independently created for each request.   I tried the shared which lead to this implementation.   There is a lot of awkwardness here as the reference counters for views and python interact.  Originally I made views embedded in the JPArray itself which other than space is more correct.  I was going to go with that solution, but the future pull request to support the numpy `__array__` is going to require creating a view object with a lot of substance as it has to open all of the views at once.  

I am thus far trying to keep as little C++ in the Python layer as possible as I am already running into serious issues with the fact that CPython is dealing with them pretty close to directly.  If you would like me to try to use smart pointers I can, but the pointer is going to get held by a C struct in PyJPArray.

Reply:
Thanks for the explanation. I can imagine that this is really tough. I think you have the right strategy here.